### ISSUE TITLE
Multiple connections

### ISSUE THREAD

**Issue by Jarogniew** [dshafik/php7-mysql-shim#39]
Due to connection hash construction it is not possible to use this library in one script which uses multiple database connections at the same time (for example with PCNTL). 

Fix:
`$hash = sha1($hostname . $username . $flags . rand(1111111, 9999999) . microtime());`

**Comment by dshafik** [comment]
@Jarogniew this would completely break the point of the hash - to reuse connections. Can you confirm that the original MySQL extension works as you expected?

**Comment by Jarogniew** [comment]
I was using for a long time native MySQL extension with PCNTL to fork separate processes from one CLI script. Using your library, when the first process has finished, the rest of them got 'MySQL server has gone' response. The workaround is to give separate MySQL link to each process is to make hash more unique. But maybe this approach will make MySQL links less efficient for normal usage, like web pages.

**Comment by staabm** [comment]
shouldn't you use non-persistent connections when you don't expect the connections to be shared?

**Comment by aalexgabi** [comment]
I got into this problem too. For me I needed to access two databases on the same server so I needed one connection per database to avoid switching database before every request.

I don't understand this part of the code:
```
            /* persistent connections start with p: */
            if ($hostname{1} !== ':' && isset(\Dshafik\MySQL::$connections[$hash])) {
                \Dshafik\MySQL::$last_connection = \Dshafik\MySQL::$connections[$hash]['conn'];
                \Dshafik\MySQL::$connections[$hash]['refcount'] += 1;
                return \Dshafik\MySQL::$connections[$hash]['conn'];
            }
```

Shouldn't this be `$hostname{1} === ':'` instead? An existing connection should only be used when 'p:' are the first two characters of the hostname according to the comment.

**Comment by aalexgabi** [comment]
@dshafik  Look at @terrafrost commit: https://github.com/terrafrost/php7-mysql-shim/commit/d0ee3daef00106f3d9381b4cc4ef378d7dedce33

And some others: https://github.com/terrafrost/php7-mysql-shim/commits/master

**Comment by techhead** [comment]
Bump. I believe that @aalexgabi is correct. Any comment?

**Comment by techhead** [comment]
> Bump. I believe that @aalexgabi is correct. Any comment?

Okay, actually not. I think that I understand the code now.

The default behavior for `mysql_connect` is to reuse the same link when called with the same arguments. This is emulated here. The code just chooses to skip the emulation when a persistent connection is used, as `mysqli_connect` will return a `mysqli` instance that represents the same underlying connection anyway. So this issue is not an issue. However, an improvement would be to actually use the ignored `$new` parameter here for better backwards compatibility with the original `mysql_connect`.

https://www.php.net/manual/en/function.mysql-connect.php
See documentation for **new_link** parameter.

```
/* persistent connections start with p: */
if (!$new && $hostname{1} !== ':' && isset(\Dshafik\MySQL::$connections[$hash])) {
		\Dshafik\MySQL::$last_connection = \Dshafik\MySQL::$connections[$hash]['conn'];
	\Dshafik\MySQL::$connections[$hash]['refcount'] += 1;
	return \Dshafik\MySQL::$connections[$hash]['conn'];
}
```

**Comment by techhead** [comment]
#45 Added pull request for suggested compatibility fix.

**Comment by dshafik** [comment]
This is now fixed with the merge of #45, thanks @techhead !
